security/s2a/internal/crypter/aesgcm_test.go

```diff
@@ -0,0 +1,474 @@
+
+	// Encrypt.
+	var dst []byte
+	if test.AllocateDst {
+		dst = make([]byte, len(test.Plaintext)+sender.TagSize())
+	}
+	got, err := sender.Encrypt(dst[:0], test.Plaintext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, ciphertext) {
+		t.Errorf("key=%v\nEncrypt(\n dst = %v\n plaintext = %v\n nonce = %v\n aad = %v\n) = (\n %v\n %v\n) want: %v",
+			test.Key, dst[:0], test.Plaintext, test.Nonce, test.Aad, got, err, ciphertext)
+	}
+
+	// Decrypt.
+	got, err = receiver.Decrypt(nil, ciphertext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, test.Plaintext) {
+		t.Errorf("key=%v\nDecrypt(\n dst = nil\n ciphertext = %v\n nonce = %v\n aad = %v\n) = (\n %v\n %v\n) want: %v",
+			test.Key, ciphertext, test.Nonce, test.Aad, got, err, test.Plaintext)
+	}
+}
+
+func testGCMEncryptRoundtrip(sender S2AAeadCrypter, receiver S2AAeadCrypter, t *testing.T) {
+	// Construct a dummy nonce.
+	nonce := make([]byte, nonceSize)
+
+	// Encrypt.
+	const plaintext = "This is plaintext."
+	var err error
+	// Reuse `buf` as both the input and output buffer. This is required to test
+	// the case where the input and output buffers fully overlap.
+	buf := []byte(plaintext)
+	ciphertext, err := sender.Encrypt(buf[:0], buf, nonce, nil)
+	if err != nil {
+		t.Fatalf("Encrypt(%v, %v, %v, nil) failed, err = %v", buf[:0], buf, nonce, err)
+	}
+
+	// Decrypt first message.
+	decryptedPlaintext, err := receiver.Decrypt(ciphertext[:0], ciphertext, nonce, nil)
+	if err != nil {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) failed, err = %v", ciphertext[:0], ciphertext, nonce, err)
+	}
+	if string(decryptedPlaintext) != plaintext {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) = %v, want = %v", ciphertext[:0], ciphertext, nonce, decryptedPlaintext, plaintext)
+	}
+}
+
+// Test encrypt and decrypt using an invalid key size.
+func TestAESGCMInvalidKeySize(t *testing.T) {
+	// Use 17 bytes, which is invalid
+	key := make([]byte, 17)
+	if _, err := NewAESGCM(key); err == nil {
+		t.Error("expected an error when using invalid key size")
+	}
+}
+
+// Test update key for AES-GCM using a key with different size from the initial
+// key.
+func TestAESGCMKeySizeUpdate(t *testing.T) {
+	for _, tc := range []struct {
+		desc          string
+		updateKeySize int
+	}{
+		{"mismatch key size update", aes256GcmKeySize},
+		{"invalid key size update", 17},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			key := make([]byte, aes128GcmKeySize)
+			crypter, err := NewAESGCM(key)
+			if err != nil {
+				t.Fatalf("NewAESGCM(keySize=%v) failed, err: %v", aes128GcmKeySize, err)
+			}
+
+			// Update the key with a new one which is a different from the original.
+			newKey := make([]byte, tc.updateKeySize)
+			if err = crypter.UpdateKey(newKey); err == nil {
+				t.Fatal("UpdateKey should fail with invalid key size error")
+			}
+		})
+	}
+}
+
+// Test Encrypt/Decrypt using an invalid nonce size.
+func TestAESGCMEncryptDecryptInvalidNonce(t *testing.T) {
```

Comment:
The `TestAESGCMEncrypt` below seems to attempt to encrypt and decrypt successfully. Can we merge this test into the one below? Basically add a new test case with an invalid nonce.

Reply:
Done